src/libraries/System.Private.CoreLib/src/System/Buffers/TlsOverPerCoreLockedStacksArrayPool.cs

```diff
@@ -150,73 +131,58 @@ public override T[] Rent(int minimumLength)
             }
-
             return buffer;
         }
 
         public override void Return(T[] array, bool clearArray = false)
         {
-            if (array == null)
+            if (array is null)
             {
-                throw new ArgumentNullException(nameof(array));
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
             }
 
             // Determine with what bucket this array length is associated
             int bucketIndex = Utilities.SelectBucketIndex(array.Length);
 
-            // If we can tell that the buffer was allocated (or empty), drop it. Otherwise, check if we have space in the pool.
+            // Make sure our TLS buckets are initialized.  Technically we could avoid doing
+            // this if the array being returned is erroneous or too large for the pool, but the
+            // former condition is an error we don't need to optimize for, and the latter is incredibly
+            // rare, given a max size of 1B elements.
+            T[]?[] tlsBuckets = t_tlsBuckets ?? InitializeTlsBuckets();
+
+            bool haveBucket = false;
             bool returned = true;
-            bool haveBucket = bucketIndex < _buckets.Length;
-            if (haveBucket)
+            if ((uint)bucketIndex < (uint)tlsBuckets.Length)
             {
-                // Clear the array if the user requests.
+                haveBucket = true;
+
+                // Clear the array if the user requested it.
                 if (clearArray)
                 {
                     Array.Clear(array);
                 }
 
-                // Check to see if the buffer is the correct size for this bucket
-                if (array.Length != _bucketArraySizes[bucketIndex])
+                // Check to see if the buffer is the correct size for this bucket.
+                if (array.Length != Utilities.GetMaxSizeForBucket(bucketIndex))
                 {
                     throw new ArgumentException(SR.ArgumentException_BufferNotFromPool, nameof(array));
                 }
 
-                // Write through the TLS bucket.  If there weren't any buckets, create them
-                // and store this array into it.  If there were, store this into it, and
-                // if there was a previous one there, push that to the global stack.  This
-                // helps to keep LIFO access such that the most recently pushed stack will
-                // be in TLS and the first to be popped next.
-                T[]?[]? tlsBuckets = t_tlsBuckets;
-                if (tlsBuckets == null)
+                // Store the array into the TLS bucket.  If there's already an array in it,
+                // push that array down into the per-core stacks, preferring to keep the latest
+                // one in TLS for better locality.
+                T[]? prev = tlsBuckets[bucketIndex];
+                tlsBuckets[bucketIndex] = array;
+                if (prev is not null)
```

Comment:
Reviewing this PR reminded me again that I find `is not null` less readable than `!= null`. I guess that's just me.